docs/proposals/23-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,354 @@
+  HVPA's policy for early horizontal scale-out.
+- Allow usage of VPA for right-sizing the resource requests during the entire time and not only under certain conditions.
+- Enable future migration away from HVPA by removing all dependency on it for scaling ShootKapi.
+- Reduce Gardener custom code base and ongoing maintenance effort.
+
+### Non-Goals
+There are other components currently scaled via HVPA. Changing the autoscaling mode of these, and of `kube-apiserver`s
+other than the one in the shoot control plane, is outside the scope of this GEP. These workloads are subject to a
+different mode of autoscaling (e.g. vertical-only, where HVPA is only used to stabilize VPA behavior), and it is
+advantageous from project execution perspective to address them as a separate concern.
+
+## Proposal
+ShootKapi HVPA is replaced by individual HPA and VPA. Undesirable interference between HPA and VPA is avoided through
+the use of sufficiently independent driving signals. HPA is driven by the rate of HTTP requests to kube-apiserver. VPA is
+driven by CPU and memory usage metrics.
+
+### Rationale
+A rough, heuristic horizontal sizing sets the stage for optimal VPA operation (horizontal size is within stable and
+reasonably efficient range). Vertical scaling fine-tunes requests, ensuring high efficiency (requests not too high)
+and stability (requests not too low).
+
+### Design Outline
+One HPA and one VPA resources are deployed alongside each ShootKapi deployment object. The VPA controls CPU and memory
+requests but not limits. The HPA is driven by the average rate of API requests per pod of the target kube-apiserver
+deployment.
+
+A new component named gardener-custom-metrics is deployed on each seed. It directly scrapes metrics data from all
+ShootKapi pods on the seed and derives custom metrics based on it. The component is registered as an extension API
+service to the seed kube-apiserver. It occupies the custom metrics API extension point and is responsible for providing
+all custom metrics for the seed kube-apiserver, including the one driving the ShootKapi HPA.
+
+![Design outline](./assets/gep23-design-outline.png "Design outline")
+
+_Fig 2: Proposed design_
+
+### Element: Gardener Custom Metrics Provider Component 
+A new component, named gardener-custom-metrics is added to seed clusters. It periodically scrapes the metrics endpoints
+of all ShootKapi pods on the seed. It implements the K8s custom metrics API and provides K8s metrics specific to
+Gardener, based on custom calculations. The proposed design can naturally be extended to multiple sources of
+input data and a non-volatile cache for acquired data. However, in the scope of this proposal, the only data source
+is the aforementioned metrics scrape, and the calculated values only need to be briefly cached in memory.
+
+A single replica is deployed, as highly available autoscaling setup is currently prevented by limitations in
+the autoscalers utilized by Gardener.
```

Comment:
I don't understand your point. It is not scope of this GEP to analyze the value of an HA setup. We did this before.
From my perspective HA for ShootKapi autoscaling is required in regards of the [HA Control Plane story](https://github.com/gardener/gardener/blob/master/docs/proposals/20-ha-control-planes.md). We are talking about a new development and it is not helpful for this to introduce yet another component which is not HA ready.
If operators want to scale down `custom-metrics-provider` in their landscape for whatever reason, they are still free to do so.

The "hypothetical risk" is not only an AZ failure, it could be node failures or a networking issues too which makes a `custom-metrics-provider` instance unavailable. In an active/passive setup it is a matter of seconds to recover, in a single replica setup it is a matter of minutes.
Even when VPA does not support an HA setup yet, it is still reducing the risk of an outage.

Reply:
>component which is not HA ready
"Not HA ready" is an overstatement. The component is very much HA ready. One can run as many replicas as they like. And active replicas at that, which is the golden standard wrt availability/reliability. It does not support active/passive mode, which is an optimisation matter. But that's exactly what I'm asking - would active/passive actually be an optimisation over single replica? If we had active/passive support, would we put it in use tomorrow, since it costs money and still does not achieve HA autoscaling? Because of VPA's vulnerability, that (figuratively speaking) 1% of availability we'd be getting, is a change between (figuratively) 98% and 99%. Not between 99% and 99.999%. Much less valuable.

Why do you say that the recovery time of a single replica is minutes? The metrics provider reaches a state equivalent to a passive replica in milliseconds, plus the time to launch the pod. Is the minutes estimate because of the potential need to scale the cluster? 

On a separate note, a passive replica would start providing data in no less than one minute. The current metric is a rate, so it only becomes available after the second scrape. Also, the natural temporal resolution of autoscaling is about 1-2 minutes. For example, under typical operation VPA makes itself "unavailable" for one minute at a time. Less lag is always better, because lag adds up. But from pragmatic perspective, 20 seconds make very little difference.

Reply:
It seems like @oliver-goetz @rfranzke and me don't understand (or agree with) your points and strongly suggest running a second instance of the new custom metrics component for the mentioned reasons.
Can you please adapt the GEP accordingly for the sake of moving forward?

Reply:
Please let us not start discussing HA back at square one.

In an active/passive setup the passive node can take over after `lease duration` + `time passive pod needs to get ready`.

When you face an unavailable node in the single replica setup it is:
`time kube-controller-manager needs/waits to set a node to status failed` + `pod-eviction-timeout` + `sometimes time for starting a new node` +  `starting the pod` + `time pod needs to get ready` 

How could the difference be just 20 seconds? The default `pod-eviction-timeout` alone is 5 minutes. Starting new nodes takes a couple of minutes too.

On the opposite side it is not a big effort to include leader election to a `controller-runtime` based controller and then operators can switch it on and off as they wish.

Reply:
I agree to not discuss HA here and stick to how we do it for all other components. I saw no good reason to deviate. Even if, that should be another discussion.

Regardless, I cannot even follow the arguments for non-HA:
- When a zone goes down and the proverbial shit hits the fan, exactly then we might need auto-scaling the most and wouldn't have it (to drive the resources in the healthy remaining zones that must take over the load)
- While the pod eviction time was reduced for our components to 60s I believe, i.e. things may recover faster in theory, they may not in reality, because if a node must be provisioned first, as everything starts growing/exploding in the remaining healthy zones, you may not get it for very long, because in reality we see resources crunches in the remaining healthy zones when one zone goes down (everybody wants to spin up something there). I.e. we cannot expect to get all machines we want, if components aren't already running.
- HA isn't driven by costs or else we wouldn't do it - it's driven by the theoretical possibility (=risk) of a zone outage (and it's also a checkbox feature, but that's something else).